### PR TITLE
Binary interpreter API improvements

### PR DESCRIPTION
**Pull request by brendanzab**
This changes the binary interpreter's API to be a bit easier to work with. Instead of supplying a module when reading an item from a `ReadScope`, the module is loaded up when the interpreter context is initially constructed. The `ReadScope` is then passed as an argument to `Context::read_item`.

### CODE REVIEW

File: fathom/src/lang/core/binary/read.rs [hunk 1]
```diff
@@ -13,57 +13,22 @@ pub struct Context<'me> {
...
     items: HashMap<String, semantics::Item>,
     locals: core::Locals<Arc<Value>>,
-    reader: FormatReader<'me>,
 }
 
 impl<'me> Context<'me> {
     /// Create a new context.
-    pub fn new(globals: &'me Globals, reader: FormatReader<'me>) -> Context<'me> {
-        Context {
+    pub fn new(globals: &'me Globals, module: &Module) -> Context<'me> {
```
Comment (toothbrush) [inline fathom/src/lang/core/binary/read.rs:20]: Yes, cool, i think this makes sense.  Will simplify the CLI code indeed.

**Comment by toothbrush** [comment]
Incidentally, that's a lot of changes!  Did your editor help you out, or did you have to go and change all of that stuff manually?

**Review by toothbrush — APPROVED** [review]

**Comment by brendanzab** [comment]
Yeah I lean a bunch on multiple cursors and find and replace and such. Can be a bit of a double edged sword though as it makes me accept a bit of verbosity early on. Really think we could have some way of making the binary tests easier with some sort of common setup stuff, but I haven't got around to it yet!

**Comment by toothbrush** [comment]
> Yeah I lean a bunch on multiple cursors and find and replace and such. Can be a bit of a double edged sword though as it makes me accept a bit of verbosity early on. Really think we could have some way of making the binary tests easier with some sort of common setup stuff, but I haven't got around to it yet!

Yep interesting.  I just wanted to know whether i was missing out on some new updated LSP Rust-specific magic or anything.  I'm "glad" to hear that's how you do it :wink:

**Comment by brendanzab** [comment]
Refactoring support is getting loads better in recent updates to the rust-analyzer at least - I'm finding I can rely lots more on smart renaming etc. Haven't tried out the other stuff though.

**Comment by wezm** [comment]
> > Yeah I lean a bunch on multiple cursors and find and replace and such. Can be a bit of a double edged sword though as it makes me accept a bit of verbosity early on. Really think we could have some way of making the binary tests easier with some sort of common setup stuff, but I haven't got around to it yet!
> 
> Yep interesting. I just wanted to know whether i was missing out on some new updated LSP Rust-specific magic or anything. I'm "glad" to hear that's how you do it 😉

I try not be too annoying about about this but IntelliJ + Rust plugin (or CLion + Rust) have _excellent_ refactoring tools. The recent renames I did in https://github.com/yeslogic/prince/pull/773 were a few seconds each with perfect accuracy (including comments, and code examples in doc comments).

**Comment by toothbrush** [comment]
> IntelliJ + Rust plugin (or CLion + Rust)

Out of curiosity, these tools aren't Libre/FOSS, right?

**Comment by wezm** [comment]
CLion isn't but Intellj CE is:

- [IntelliJ CE](https://github.com/JetBrains/intellij-community) (Apache 2)
- [IntelliJ Rust plugin](https://github.com/intellij-rust/intellij-rust) (MIT)

**Comment by toothbrush** [comment]
Cool, great news!

On 2020-12-09 at 12:25 ACST+0930, quoth Wesley Moore 
<notifications@github.com>:
> CLion isn't but Intellj CE is:
>
> - [IntelliJ CE](https://github.com/JetBrains/intellij-community) 
> (Apache 2)
> - [IntelliJ Rust 
> plugin](https://github.com/intellij-rust/intellij-rust) (MIT)

**Pull request merged**